(Issue #41) Add Support for Oauth2 authorization header / bearer token

tato123 opened the pull request:
Added support for the Oauth2 authorization header from RFC 6750, allows users to send any HTTP verb with a Oauth2 compliant bearer token.

```
GET /resource HTTP/1.1
Host: server.example.com
Authorization: Bearer mF_9.B5f-4.1JqM
```

Included corresponding test cases

1. Backwards compatibility test for access_token header
2. Authorization header test
3. Lowercase test (some clients will transform http headers to all lowercase like unirest)

Updated readme documentation to better highlight all of the available mechanisms for clients (query param, body, and headers (both custom and authorization)

ghaiklor commented on .gitignore:
```diff
@@ -26,7 +26,7 @@ build/Release
 
 # Babel compiled sources
-lib
+#lib
```
Why did you comment this? Compiled sources shouldn't be available on GitHub.

tato123 replied:
Sorry about that, didn't mean to include that as part of the pull request. 

I was attempting to make sure post build and pull down from git that things worked as expected (since I don't have an npm account to publish to and didn't want to publish a fork of this project on npm)

I'll revert this change.

ghaiklor commented on src/index.js:
```diff
@@ -51,9 +51,9 @@ export default class FacebookTokenStrategy extends OAuth2Strategy {
    */
   authenticate(req, options) {
-    let accessToken = (req.body && req.body[this._accessTokenField]) || (req.query && req.query[this._accessTokenField]) || (req.headers && req.headers[this._accessTokenField]);
+    let accessToken = (req.body && req.body[this._accessTokenField]) || (req.query && req.query[this._accessTokenField]) || (FacebookTokenStrategy.parseAccessTokenHeader(req.headers, this._accessTokenField, this._authorizationField));
```
If you made static method that parses access token then better to move there body and query too. Like `let accessToken = FacebookTokenStrategy.parseAccessToken(req)`. What you think?

tato123 replied:
I agree with your suggestion, it was my original thought as well but I wanted to avoid making too many changes in the library (plus I'm not as familiar with ES6 yet and wasn't sure if there was a better way to handle things).

I will update it.

tato123 commented:
Updated pull request
 - reverted .gitignore change to include compiled lib folder
 - modified the static parseAccessToken to include support for query and body parsing

ghaiklor commented on src/index.js:
```diff
@@ -51,9 +51,9 @@ export default class FacebookTokenStrategy extends OAuth2Strategy {
    */
   authenticate(req, options) {
-    let accessToken = (req.body && req.body[this._accessTokenField]) || (req.query && req.query[this._accessTokenField]) || (req.headers && req.headers[this._accessTokenField]);
+    let accessToken =  (FacebookTokenStrategy.parseAccessToken(req, this._accessTokenField, this._authorizationField));
```
You can implement it more even simpler. Kind of:

```js
let accessToken = FacebookTokenStrategy.lookup(req, this._accessTokenField);

lookup = (req, field) {
  return (
    req.body && req.body[field] ||
    req.query && req.query[field] ||
    req.headers && req.headers[field]
  );
}
```

Then you don't need to implement methods for parse `accessToken` or `refreshToken` and just call method

```js
let accessToken = FacebookTokenStrategy.lookup(req, this._accessTokenField)
let refreshToken = FacebookTokenStrategy.lookup(req, this._refreshTokenField);
```

tato123 replied:
I'm not as familiar with ES6 so I'm not quite understanding how to implement your suggestion. The suggested change doesn't seem to handle the OAuth2 Authorization http header from the OAuth2 spec or parsing the Bearer token (which was what I was trying to add as part of this pull request)

ghaiklor replied:
@tato123 just add your parsing login into lookup method as new OR condition. Like `req.headers && req.headers[field] || yourLogicToParse()`.

tato123 commented:
Updated based on your feedback:
 - Refactored to a lookup method that used more compact syntax
 - Updated refresh token to use use lookup method
 - Split OAuth2 token parsing (refactored to compact statements)

One thing to note, I retained the lowercase matching to perform header matching compliant with RFC 2616 section 4.2 spec.
